### ISSUE TITLE
[Feature Request]: "streaming reading" from web server

### ISSUE THREAD

**Issue by GeistPrisma** [ollm/OpenComic#260]
### Preflight Checklist

- [X] I have searched the [issue tracker](https://www.github.com/ollm/OpenComic/issues) for a feature request that matches the one I want to file, without success.

### Describe your suggested feature

To read the comic from web server, OpenComic need to download the full file. Most manga files are in cbz format, which is the same format as zip. Due to the unique structure of ZIP file, it support reading the contained files even if the download is incomplete. Therefore, if the files are stored on a server that supports partial downloads, such as WebDAV or SMB, we can achieve streaming reading by parsing the structure of the ZIP file, without needing to download the entire file.

### Additional Information

**zip structure** https://en.wikipedia.org/wiki/ZIP_(file_format)#Structure
**stream unzip** https://github.com/ShiinaRinne/ZipStreamLoader
https://github.com/mhr3/unzip-stream

### Sample file

I have yet to find a comic reader on Windows that supports this feature. The KedaReader on the App Store and the WoWo Comic Reader on google play store have implemented this feature very well.

**Comment by ollm** [comment]
This probably can't be done, `zip` file format isn't really meant to be processed by streaming ([unzip-stream#note](https://github.com/mhr3/unzip-stream?tab=readme-ov-file#unzip-stream-)), and OpenComic also supports more types of compressed files/servers, that would cause more difficulties in its implementation. Also it would be incompatible with some features, currently when a file has been downloaded for reading, it is temporarily saved for future reading in `tmp` folder, and is not downloaded again, It is only deleted when the size set from the settings is exceeded (4GB by default). Because of this, there is also an offline reading feature that attempts to read from temporary files when there is no connection to the server.

Due this and other reasons, I have no plans to implement this feature.

> Now "webdav://" connects to the server correctly. However, when switching to Library label or serve label, OpenComic will auto download all files from the webdav server. "smb://" has the same problem ,the other serves I haven't tested yet. 
> _Originally posted by @GeistPrisma in https://github.com/ollm/OpenComic/issues/239#issuecomment-2295118473_
 
About this, it's a bug, OpenComic downloads the files to generate the thumbnails, but it wasn't reading the cached data correctly afterwards, downloading them again every time, as far as I could see, it only happened in compressed files, I push this fix 7e739dae8cbcef9691efd0a0cfbfbfa62ce192a4, but if you detect some repetitive behavior, it's possible that there is still some other bugs.

Build: https://mega.nz/file/bSgCGRiQ#aHDL5Q8rmEBTfhbuy0pqH0sjorlSPs31A9iquEul9W8

Additionally, if you want to speed up more the reading of your server by OpenComic, you can do the following:

**Structure in folders/subfolders**
If you structure your server in folders and subfolders, OpenComic will only download the first 4 files in each folder to generate the thumbnails.

**Have the files already extracted**
If you already have the files extracted, OpenComic will download only the image needed for the thumbnail.

Although it is likely that with just the fix it will work much better for you.

**Comment by sud0xenn** [comment]
@ollm I'm lucky to finally find a reader for windows with webdav support, tysm!! :D

Without reopening the issue, is it possible for `.cbr` formats? Live streaming (loading directly in memory) without ever writing to disk? That was the entire point for me to setup a webdav server for my comic on my phone, to not download and waste space again on my laptop when the comics are already stored on my phone.

Also, I just set up a webdav server but it does not let me open any comics:

<img width="1599" height="866" alt="Image" src="https://github.com/user-attachments/assets/70432b46-d7df-4971-81d0-3de6b1518d7f" />

**Comment by ollm** [comment]
Sorry, but OpenComic is not designed to read/extract compressed files directly in memory (and the library used for reading/extracting files doesn't even support this, [`node-7z`](https://www.npmjs.com/package/node-7z) and [`7zip-bin-full`](https://www.npmjs.com/package/7zip-bin-full)). If you're concerned about files taking up too much space, you can reduce the maximum size of temporary files in the settings, although it's always recommended not to set it to 0, since files will be read from there if they've already been downloaded/extracted previously.

Also, the latest version includes improvements to reduce file downloads/reads when generating thumbnails https://github.com/ollm/OpenComic/issues/366#issuecomment-3073635715

Regarding the issue you're experiencing, you can try the latest master build to see if it has already been fixed.

Build: https://mega.nz/file/OWQxjD6T#ppsxULzVjTeRhlqzYCk9VzIWEiz0GPzMbq9i1XjsKok

If it's not fixed, you can check the console for any error messages: `Debug` > `Toggle DevTools` and tell me what server client you are using.

**Comment by sud0xenn** [comment]
> Sorry, but OpenComic is not designed to read/extract compressed files directly in memory (and the library used for reading/extracting files doesn't even support this, [`node-7z`](https://www.npmjs.com/package/node-7z) and [`7zip-bin-full`](https://www.npmjs.com/package/7zip-bin-full)).

If those libraries start supporting it, then it would be no hassle to have that feature here too right?

> If it's not fixed, you can check the console for any error messages: `Debug` > `Toggle DevTools` and tell me what server client you are using.

I tried the latest build you mentioned, same problem. I'm running `wsgidav` (installed via pip) from termux on my phone.
The Console shows these errors:

<img width="1919" height="195" alt="Image" src="https://github.com/user-attachments/assets/c94ef7e1-0250-4160-aeb2-a1a8c3eaba9d" />

The directory `C:\Users\<user>\AppData\Local\Temp\opencomic\` contains empty folders for comics I try to open.

**Comment by ollm** [comment]
> If those libraries start supporting it, then it would be no hassle to have that feature here too right?

They would probably have to make several changes, since currently OpenComic is optimized to save the most used files in the Temp folder so that they don't need to be extracted/downloaded multiple times.

> I tried the latest build you mentioned, same problem. I'm running wsgidav (installed via pip) from termux on my phone.
The Console shows these errors:

The error seems to be caused by the colon in the port (it's not allowed as a file name in Windows). Can you share a screenshot of how you have the WebDav server path configured in the settings?, since that part of the path shouldn't appear as a folder

**Comment by sud0xenn** [comment]
> Can you share a screenshot of how you have the WebDav server path configured in the settings?, since that part of the path shouldn't appear as a folder

<img width="1080" height="824" alt="Image" src="https://github.com/user-attachments/assets/a5edb978-e29c-4be8-81c7-ecd065aeea56" />

<img width="888" height="147" alt="Image" src="https://github.com/user-attachments/assets/b2096e99-c21b-408b-89b6-25583de8b433" />

**Comment by ollm** [comment]
I've found the main cause of the bug, this should fix it 65ecdf6e5ef94170cce3ebcf0062fd2b3ebac2fe

If there is still any other error let me know.

Build: https://mega.nz/file/7exwnAQL#deJQRWvQ0K7TBd9Aup32FSh2RWtXxqhKKvnag1VCNUA

**Comment by sud0xenn** [comment]
> I've found the main cause of the bug, this should fix it [65ecdf6](https://github.com/ollm/OpenComic/commit/65ecdf6e5ef94170cce3ebcf0062fd2b3ebac2fe)
> 
> If there is still any other error let me know.
> 
> Build: https://mega.nz/file/7exwnAQL#deJQRWvQ0K7TBd9Aup32FSh2RWtXxqhKKvnag1VCNUA

That was quick! Testing it out, will edit this comment :D

@ollm It works now thank you!!
Although performance wise a bit laggy on the loading part. I really wish there existed a webdav streaming solution for self hosting. Anyway going to test opencomic for sometime, thanks again.

**Comment by sud0xenn** [comment]
For some files I'm getting this error now:

<img width="623" height="429" alt="Image" src="https://github.com/user-attachments/assets/e1c9b0fd-fb4e-4163-878c-8c9782be5007" />

<img width="838" height="266" alt="Image" src="https://github.com/user-attachments/assets/28c5d7d1-d6bc-42b7-bf01-1b4e4d433330" />

EDIT: Just noticed they are `.cbz` files which are getting this error
EDIT2: But I also have other `.cbz` files which open perfectly fine

**Comment by ollm** [comment]
I'm not entirely sure, but this should fix this error a44a43e7ce08c8e9a5fa9f2150678c4a2f370398

Build: https://mega.nz/file/DbxmyBzR#uCOoF4S1LjE9F3lUXXwLNs6dPZm3F1VaCaj13Zx8PbI

**Comment by sud0xenn** [comment]
> I'm not entirely sure, but this should fix this error [a44a43e](https://github.com/ollm/OpenComic/commit/a44a43e7ce08c8e9a5fa9f2150678c4a2f370398)
> 
> Build: https://mega.nz/file/DbxmyBzR#uCOoF4S1LjE9F3lUXXwLNs6dPZm3F1VaCaj13Zx8PbI

Tried the new build, exact same error as above.

**Comment by ollm** [comment]
Can you try with the CBZ files locally, without using WebDav? That way we can see if it's an issue with the files or with WebDav.

**Comment by sud0xenn** [comment]
> Can you try with the CBZ files locally, without using WebDav? That way we can see if it's an issue with the files or with WebDav.

Local files work fine, same one tested here which was bugging in webdav:
<img width="1920" height="1020" alt="Image" src="https://github.com/user-attachments/assets/fc6e22cb-e6ff-486c-8b54-c16efb34a0c8" />

**Comment by ollm** [comment]
I've made some more changes to see if the problem gets fixed cf3ea7c3a61fdf8070ef1da6aa965e22390b15e1

Build: https://mega.nz/file/yXIhmbAY#XqYrf2rSXTNedR433QmfXBQWo5oC5VNG_IJzdjB1-UQ

**Comment by sud0xenn** [comment]
> Build: https://mega.nz/file/yXIhmbAY#XqYrf2rSXTNedR433QmfXBQWo5oC5VNG_IJzdjB1-UQ

The issue persists.
Here's an extra error tested on another cbz file:

<img width="650" height="198" alt="Image" src="https://github.com/user-attachments/assets/49935c69-1183-4540-b1eb-539e643034f3" />

I think the issue here again is unsanitized input in filename. The `,` is likely the culprit. In the previous file above, maybe it was the `-`.

**Comment by ollm** [comment]
> <img alt="Image" width="650" height="198" src="https://private-user-images.githubusercontent.com/226022295/479636833-49935c69-1183-4540-b1eb-539e643034f3.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU4NjgzNTcsIm5iZiI6MTc1NTg2ODA1NywicGF0aCI6Ii8yMjYwMjIyOTUvNDc5NjM2ODMzLTQ5OTM1YzY5LTExODMtNDU0MC1iMWViLTUzOWU2NDMwMzRmMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyMlQxMzA3MzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMTQ4MjYyNzcwNWI5NWQxYWUxOGUzMjQ3MzhiNzVkZTZjOTc0MzZmNGYxYTdkMjdlY2IwNTEyMmZlNGU0NTUxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.RtNKCL37ntQd3C-_-nBVcvbezgNherapM23POnze__E">
> I think the issue here again is unsanitized input in filename. The `,` is likely the culprit. In the previous file above, maybe it was the `-`.

At first, these characters shouldn't cause any issues, since they are allowed characters in Windows.

I've fixed some errors I found when extracting files e4712cee2ced2176cb666e772cdf4c88f30a6b3f, but I don't think this will solve the problem, as I believe there's an issue with downloading the files from the server in OpenComic. For now, my recommendation is to try deleting the temporary files (in OpenComic settings, at the very bottom) to see if they download correctly again.

Build: https://mega.nz/file/mH5RVbCY#LJwy6n3bzahJXAOKbZSwk46iIrXKM7mblWffwpUxGAA

**Comment by sud0xenn** [comment]
> but I don't think this will solve the problem

Yea, didn't fix it. Do you think this is a bug in the unzipping libraries used? Or maybe my CBZ files have missing expected files

**Comment by ollm** [comment]
> Yea, didn't fix it. Do you think this is a bug in the unzipping libraries used? Or maybe my CBZ files have missing expected files

I don't think it has to do with the unzipping libraries or with your cbz files, since they open fine locally.

Most likely there's some error in OpenComic logic that I can't manage to find, either in the server conexion part or in how compressed files are handled (or in both at the same time).

I've made a build that prints more error messages to the console, so we can know a bit better where it happens. Now something similar to this should appear.

<img width="582" height="80" alt="Image" src="https://github.com/user-attachments/assets/2875f294-ece8-4c9b-a78c-7423219b9456" />

Build: https://mega.nz/file/zWplwb5L#RtjxfQGH_CZ9XwATeYOhM0RnT3ZwM9iiSSc_6jG4Tb4

**Comment by sud0xenn** [comment]
> Build: https://mega.nz/file/zWplwb5L#RtjxfQGH_CZ9XwATeYOhM0RnT3ZwM9iiSSc_6jG4Tb4

<img width="1270" height="861" alt="Image" src="https://github.com/user-attachments/assets/cd361d31-4ebd-4939-84fa-3faf408001f6" />

index.html:
```
dom.loadIndexPage(true, '.\\webdav:\\myphone.local:8080\\Marvel Adventures Spider-Man(2005-2010)\\Marvel Adventures Spider-Man 001 (2005) (Digital) (Shadowcat-Empire).cbz', false, false, '.\\webdav:\\myphone.local:8080\\Marvel Adventures Spider-Man(2005-2010)');
```



Also, I usually see that yellow error. Maybe increase the cooldown between timers?

**Comment by ollm** [comment]
Finally, I was able to reproduce the problem. There was an error when retrieving the list of files from the subfolders of the compressed file. I've submitted this fix b1c7aef929666ebe0f8b5283cdb8bde19051ba91

The yellow warnings aren't significant. They occur because each file download starts a `console.time()` to measure its duration. If another console.time() with the same label is triggered before the previous one ends with `console.timeEnd()`, the console shows this warning.

Build: https://mega.nz/file/SLADTY5R#_GyyxtdurU6UtNFf8YcGloh2SPyhswQoh0LHLrbw7J0
